feat(core): Optionally ignore missing env vars during expansion

WillDaSilva opened the pull request:
In addition to addressing #7122, this also includes some performance improvements that have a significant effect when using `expand_env_vars` on many large env dictionaries, as will be done by the implementation for [Meltano manifest](https://github.com/meltano/meltano/issues/6876).

Some work to wrangle the complexity of `expand_env_vars` was made, but it's widely used by legacy code, which makes it difficult to improve significantly.

For instance, `expand_env_var` could previous be called on any type, and it would return it. This is now disallowed by the type hint, but still supported in the code itself. Hopefully this will discourage these bad practices.

Another questionable aspect of `expand_env_vars` that had to be preserved is that it will perform a recursive tree-traversal on any mapping structure it is provided as `raw_value`. This dangerous/slow/complex behaviour can now be opted out of by setting the `flat` paramter to `True`, which asserts that the `raw_value` being provided has a flat structure (i.e. no nested dicts).

Closes #7122

WillDaSilva committed:
feat(core): Optionally ignore missing env vars during expansion

In addition to addressing #7122, this also includes some performance improvements that have a significant effect when using `expand_env_vars` on many large env dictionaries, as will be done by the implementation for [Meltano manifest](https://github.com/meltano/meltano/issues/6876).

Some work to wrangle the complexity of `expand_env_vars` was made, but it's widely used by legacy code, which makes it difficult to improve significantly.

For instance, `expand_env_var` could previous be called on any type, and it would return it. This is now disallowed by the type hint, but still supported in the code itself. Hopefully this will discourage these bad practices.

Another questionable aspect of `expand_env_vars` that had to be preserved is that it will perform a recursive tree-traversal on any mapping structure it is provided as `raw_value`. This dangerous/slow/complex behaviour can now be opted out of by setting the `flat` paramter to `True`, which asserts that the `raw_value` being provided has a flat structure (i.e. no nested dicts).

Closes #7122

WillDaSilva committed:
Improve `test_expand_env_vars`

WillDaSilva committed:
Merge branch 'main' into 7122-ignore-missing-env-var

edgarrmondragon commented on src/meltano/core/utils/__init__.py:
```diff
@@ -478,42 +477,111 @@ def __init__(self, env_var: str):
-            key: expand_env_vars(val, env, raise_if_missing)
-            for key, val in raw_value.items()
-        }
-    elif not isinstance(raw_value, str):
+    ignore_if_missing: bool = False,
```
wdyt about refactoring these args to use a single flag instead:

1. EmptyStr
2. Raise
3. Ignore

WillDaSilva replied:
I had thought about doing that, but didn't want to break every spot that uses the pattern:

```python
with project_settings_service.feature_flag(
    FeatureFlags.STRICT_ENV_VAR_MODE, raise_error=False
) as strict_env_var_mode:
    expand_env_vars(..., raise_if_missing=strict_env_var_mode)
```

`ProjectSettingsService.feature_flag` returns a bool which is passed in directly. We could map that bool to the appropriate enum value at every call site, but that's a bit messy.

One hacky approach we could take is to make it an int enum, have 0 correspond to the default behaviour, and have 1 correspond to raising if missing. That way a bool passed in would work without any changes. What do you think about that @edgarrmondragon?

edgarrmondragon replied:
@WillDaSilva Yeah good point, quick search reveals 16 uses of `raise_if_missing=...` 😬 . I'm ok with the hack, with a note to refactor later perhaps?

WillDaSilva replied:
@edgarrmondragon Implemented in cf2333d95a0c51a17173ca9b9fd880903ffd59fa. I'm not sure if it's an improvement or not. If you have strong feelings one way or the other we can keep or revert the changes.

edgarrmondragon replied:
@WillDaSilva I think it's a good compromise to avoid having mutually exclusive params 👍

WillDaSilva committed:
Merge branch 'main' into 7122-ignore-missing-env-var

WillDaSilva committed:
Combine `raise_if_missing` and `ignore_if_missing` into int enum

This is a bit of a hack in order to maintain compatibility with the the `strict_env_var_mode` feature flag (which is a boolean), as described here: https://github.com/meltano/meltano/pull/7123#discussion_r1054823127

WillDaSilva committed:
Merge branch 'main' into 7122-ignore-missing-env-var
